operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakController.java

```diff
@@ -18,59 +18,139 @@
+import io.javaoperatorsdk.operator.api.reconciler.ControllerConfiguration;
+import io.javaoperatorsdk.operator.api.reconciler.DeleteControl;
+import io.javaoperatorsdk.operator.api.reconciler.ErrorStatusHandler;
+import io.javaoperatorsdk.operator.api.reconciler.EventSourceContext;
+import io.javaoperatorsdk.operator.api.reconciler.EventSourceInitializer;
+import io.javaoperatorsdk.operator.api.reconciler.Reconciler;
+import io.javaoperatorsdk.operator.api.reconciler.RetryInfo;
+import io.javaoperatorsdk.operator.api.reconciler.UpdateControl;
+import io.javaoperatorsdk.operator.processing.event.ResourceID;
+import io.javaoperatorsdk.operator.processing.event.source.EventSource;
+import io.javaoperatorsdk.operator.processing.event.source.informer.InformerEventSource;
+import io.quarkus.logging.Log;
+import org.keycloak.operator.Config;
+import org.keycloak.operator.Constants;
+import org.keycloak.operator.WatchedResourcesStore;
 import org.keycloak.operator.v2alpha1.crds.Keycloak;
 import org.keycloak.operator.v2alpha1.crds.KeycloakStatus;
+import org.keycloak.operator.v2alpha1.crds.KeycloakStatusBuilder;
 
-@ControllerConfiguration(namespaces = Constants.WATCH_CURRENT_NAMESPACE, finalizerName = Constants.NO_FINALIZER)
-public class KeycloakController implements Reconciler<Keycloak> {
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
 
-    @Inject
-    Logger logger;
+import static io.javaoperatorsdk.operator.api.reconciler.Constants.NO_FINALIZER;
+import static io.javaoperatorsdk.operator.api.reconciler.Constants.WATCH_CURRENT_NAMESPACE;
+
+@ControllerConfiguration(namespaces = WATCH_CURRENT_NAMESPACE)
+public class KeycloakController implements Reconciler<Keycloak>, EventSourceInitializer<Keycloak>, ErrorStatusHandler<Keycloak> {
 
     @Inject
     KubernetesClient client;
 
+    @Inject
+    Config config;
+
+    private final WatchedResourcesStore watchedSecrets = new WatchedResourcesStore();
+
+    @Override
+    public List<EventSource> prepareEventSources(EventSourceContext<Keycloak> context) {
+        SharedIndexInformer<Deployment> deploymentInformer =
+                client.apps().deployments().inAnyNamespace()
+                        .withLabels(Constants.DEFAULT_LABELS)
+                        .runnableInformer(0);
+
+        EventSource deploymentEvent = new InformerEventSource<>(
+                deploymentInformer, d -> {
+                    List<OwnerReference> ownerReferences = d.getMetadata().getOwnerReferences();
+                    if (!ownerReferences.isEmpty()) {
+                        return Set.of(new ResourceID(ownerReferences.get(0).getName(), d.getMetadata().getNamespace()));
+                    } else {
+                        return Collections.emptySet();
+                    }
+        });
+
+        // TODO optimize this (set labels?) to not watch all namespaces
```

Comment:
We could use labels of course. But it'd add additional complexity (to already complex problem ;)) – we need to remove those labels once no longer used in any CR etc. Wondering, if the "Dependent Resources" in the SDK could help with this.

Reply:
> we need to remove those labels once no longer used in any CR etc.

why?
I think we can rely on one or more ["Recommended Labels"](https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/)

Reply:
@vmuzikar heads up since I didn't knew about, but a possible approach that we can explore is to have a Reconciler for those `Secrets`, as described in the section `Support for Well Known (non-custom) Kubernetes Resources` in https://javaoperatorsdk.io/docs/features